### PR TITLE
[FrameworkBundle] Allow setting private services with the test container

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

This PR allows mocking a service for a test case. A typical example could be for HttpClient:

```php
static::getContainer()->set('example.client', new MockHttpClient(...));

// do stuff that will use `example.client` and will get mock responses instead of using the network
```

### CODE REVIEW

**Comment by damienalexandre** [comment]
Really looking forward this one! Thanks a lot!

**Comment by bastnic** [comment]
Awesome!

File: src/Symfony/Bundle/FrameworkBundle/Test/TestContainer.php [hunk 1]
```diff
@@ -69,7 +68,16 @@ public function setParameter(string $name, mixed $value)
...
-        $this->getPublicContainer()->set($id, $service);
+        $renamedId = $this->renamedIds[$id] ?? $id;
+
+        try {
+            $this->getPublicContainer()->set($renamedId, $service);
+        } catch (InvalidArgumentException $e) {
+            if (!str_starts_with($e->getMessage(), "The \"$renamedId\" service is private")) {
+                throw $e;
+            }
+            $this->getPublicContainer()->privates[$renamedId] = $service;
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Test/TestContainer.php]: should we throw if the private service was already instantiated, to avoid WTF debugging about why the new instance being set is not used by other services ? This would be consistent with the error throw in Container::set when trying to replace a public service that has already been instantiated.

Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Test/TestContainer.php]: yup! added!

**Comment by Seldaek** [comment]
Is this cool? Yes I think so. It'd make tests less brittle if you don't have to instantiate smth but can rely on auto wiring too.

Just one thought: could it detect if a service which was set is then not used during a test and output a warning? That'd allow to catch errors where you configure stuff that's useless / dead code. Possibly would back fire tho and be more annoying than useful :)

**Comment by nicolas-grekas** [comment]
@Seldaek that'd be neat but I don't know how we could make that work. There is no infrastructure in the container to know when a service is injected into another one, and even less to know when such a service is actually used.

On a case by case basis, it's doable of course:

```php
$client = new TraceableHttpClient(new MockHttpClient(...));
static::getContainer()->set('example.client', $client);

// do stuff

// do expectations using TraceableHttpClient's API
```

**Comment by tgalopin** [comment]
This would also be super useful to mock the clock (I do it all the time, mocking in the config isn't enough to test all cases):

```php
static::getContainer()->set(ClockInterface::class, new MockClock('2022-11-23 00:00:00', 'UTC'));
```

**Comment by nicolas-grekas** [comment]
@tgalopin see also https://github.com/symfony/symfony/pull/48642
TL;DR, use `ClockSensitiveTrait::mockTime()`.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/TestServiceContainerRealRefPass.php [hunk 2]
```diff
@@ -29,15 +30,35 @@ public function process(ContainerBuilder $container)
...
+                if ($id !== $target) {
+                    $renamedIds[$id] = $target;
+                }
             } else {
                 unset($privateServices[$id]);
             }
         }
 
+        foreach ($container->getAliases() as $id => $target) {
+            while ($container->hasAlias($target = (string) $target)) {
```
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/TestServiceContainerRealRefPass.php]: `do {} while()` to call `hasAlias` 1 less time?

File: src/Symfony/Bundle/FrameworkBundle/Test/TestContainer.php [hunk 3]
```diff
@@ -69,7 +68,20 @@ public function setParameter(string $name, mixed $value)
...
     public function set(string $id, mixed $service)
     {
-        $this->getPublicContainer()->set($id, $service);
+        $container = $this->getPublicContainer();
+        $renamedId = $this->renamedIds[$id] ?? $id;
+
+        try {
+            $container->set($renamedId, $service);
+        } catch (InvalidArgumentException $e) {
+            if (!str_starts_with($e->getMessage(), "The \"$renamedId\" service is private")) {
```
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/Test/TestContainer.php:77]: Couldn't we use a magic number as an exception code instead to bulletproof this?

**Review by dunglas — APPROVED** [review]
Awesome!

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/TestServiceContainerRealRefPass.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/TestServiceContainerRealRefPass.php]: wouldn't work, read again ;)

File: src/Symfony/Bundle/FrameworkBundle/Test/TestContainer.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Test/TestContainer.php:77]: we do this already and this is tested so I wouldn't mind

**Review by chalasr — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Comment by Bilge** [comment]
1. How does this work given the container has already been compiled and, where such services are replaced, the object they are replacing has already been injected into dependents and thus will not see the effect of this replacement?
2. For @nicolas-grekas specifically, seeing as this PR is at odds with your [original assessment](https://github.com/symfony/symfony/issues/29109#issuecomment-436372455), what changed?

**Comment by stof** [comment]
@Bilge you must do the injection before services are instantiated.
Compiling a container is not when instantiation happens.

**Comment by stof** [comment]
And the special TestContainer allowing you to retrieve private services already has a restriction that this does not support inlined services. This injection has the same restriction.

**Comment by wslawski-printify** [comment]
Does this also allow to set services that never existed in container in first place? If yes this is actually a pretty dangaerous feature to have, you can make all your tests pass but your production code will still explode. Could be good to add some check if it already exists in container - if not - then throw exception. Just noticed this functionality now. @nicolas-grekas 

Okay checked it - and yea, it allows to set anything in container, even if it does not exist before, sounds a bit dangerous to me, I think this could be improved by some configuration option and probably on next major version throw exception in case of trying to set service that didn't exist before.
